### ISSUE TITLE
apparmor="DENIED" in syslog because 'snap.wekan.mongodb'

### ISSUE THREAD

**Issue by webenefits** [wekan/wekan#4855]
Hi,

during a routine check of the syslogs today, I noticed that wekan produces an extremely large number of entries by `apparmor="DENIED"` in the `/var/log/syslog`. The following is an excerpt:
```
Mar 14 17:27:36 kernel: [21865.662511] audit: type=1400 audit(1678811256.000:86815): apparmor="DENIED" operation="open" profile="snap.wekan.mongodb" name="/proc/13186/net/netstat" pid=13186 comm="ftdc" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
Mar 14 17:27:36 kernel: [21865.662516] audit: type=1400 audit(1678811256.000:86816): apparmor="DENIED" operation="open" profile="snap.wekan.mongodb" name="/proc/13186/net/snmp" pid=13186 comm="ftdc" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
Mar 14 17:27:36 kernel: [21865.662518] audit: type=1400 audit(1678811256.000:86817): apparmor="DENIED" operation="open" profile="snap.wekan.mongodb" name="/proc/13186/mountinfo" pid=13186 comm="ftdc" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
Mar 14 17:27:36 kernel: [21865.662519] audit: type=1400 audit(1678811256.000:86818): apparmor="DENIED" operation="open" profile="snap.wekan.mongodb" name="/proc/vmstat" pid=13186 comm="ftdc" requested_mask="r" denied_mask="r" fsuid=0 ouid=0
```

The entries are produced approx. every second, which is why the syslog becomes quite large during the day.

Of course, I have already done some research, but unfortunately I have not found any helpful information.
1. In [this thread »](https://forum.snapcraft.io/t/how-to-connect-to-localhost-mongodb-in-snap-apparmor-prevents/7793) it's mentioned that _"the apparmor denials can be solved in wekan.mongodb by connecting the ‘network-observe’ interface"_, but if I list the snap interfaces due the command `snap connections --all` there is no 'network-observe' for wekan:
   ```
   Benutzeroberfläche  Plug                Slot                Notizen
   content             -                   wekan:mongodb-slot  -
   content             wekan:mongodb-plug  -                   -
   network             wekan:network       :network            -
   network-bind        wekan:network-bind  :network-bind       -
   ```
2. [Here »](https://mummila.net/kommentit/avainsanat/apparmor/) there was a similar problem, but it seems to have been fixed by a snap update to version 2.32.1.
3. Furthermore, I have already tried to install the current beta version 6.77 of Wekan, which has not fixed the error either.
4. I have already carried out a complete system update via `apt update && apt upgrade` including a following reboot.

**Does anyone else has this problem or even a solution?**

### Server Setup Information
- OS: Ubuntu 18.04.6 LTS
- Snap: 2.58.2
- Wekan (via Snap): 6.76
- Meteor-Version: 2.10.0
- Node-Version: 14.21.3
- MongoDB-Version: 6.0.4
- HTTP frontend: Apache

**Comment by xet7** [comment]
That is a MongoDB bug. Not WeKan bug. Currently I'm In Progress of changing database to SQLite.

**Comment by mfilser** [comment]
I don't have this bug in my Docker setup. Maybe it's related to Snap ?

**Comment by xet7** [comment]
For MongoDB database logging, snap has options to where to send logs:
https://github.com/wekan/wekan/blob/master/snap-src/bin/wekan-help#L29

Anyone interested in this could search web with `snap apparmor` , and also mongodb config options, is it possible to have less logging.

**Comment by xet7** [comment]
Or search Snap Docs https://snapcraft.io/docs

**Comment by mfilser** [comment]
I just wrote a comment on this, i think this is the problem.

https://github.com/wekan/wekan/issues/4214#issuecomment-1468786269

**Comment by mfilser** [comment]
MongoDB is fine

**Comment by webenefits** [comment]
Thank you very much for your help. 

Unfortunately, the [mentioned option mongo-log-destination »](https://github.com/wekan/wekan/issues/4855#issuecomment-1468768042) did not work. No matter whether `'snapcommon'` or `'devnull'`, the errors were written to the syslog. All other values resulted in the service not being able to start.

But with the Infos from [https://snapcraft.io/docs/debug-snaps#heading--apparmor](https://snapcraft.io/docs/debug-snaps#heading--apparmor) I edited the File '/var/lib/snapd/apparmor/profiles/snap.wekan.mongodb' and added following lines at the very end:
```
@{PROC}/@{pid}/net/netstat r,
@{PROC}/@{pid}/mountinfo r,
@{PROC}/@{pid}/net/snmp r,
/proc/vmstat r,
```

Then, to compile, load the policy into the kernel and restart the service:
```
sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.wekan.mongodb
sudo systemctl restart snap.wekan.mongodb.service snap.wekan.wekan.service
```

That seems to have fixed the problem. :tada:

However, as I did not understand 100% what exactly the error message originally said, it would be great if someone could check whether this is ok, especially with regard to security.

**Comment by xet7** [comment]
@webenefits 

Just quessing: Does MongoDB server include some free monitoring, that could be disabled with some database command or CLI option? If that monitoring tries to connect to MongoDB company provided monitoring service, but apparmor denies it? Because when I login to MongoDB serveri with MongoDB shell mongosh, there are some messages about monitoring.

**Comment by xet7** [comment]
@webenefits 

If that monitoring exists, you could try those database commands to turn monitoring off, remove you apparmor settings, and see if there is any difference.

**Comment by webenefits** [comment]
Hi @xet7 

to be honest, I can't really answer that question. We don't use MongoDB otherwise. I think only Wekan uses it and so Wekan installed MongoDB initially. Unfortunately, I don't really know much about it either, so I can't give you much more information.

If it helps you, you can send me commands and I'll be happy to provide you with the output.

**Comment by xet7** [comment]
@webenefits 

If you are logging into mongosh shell, type these commands:
```
disableTelemetry()

db.disableFreeMonitoring()
```
If you have older MongoDB shell and bash, you can also use bash script these commands:
```
mongo --eval "disableTelemetry()"

mongo --eval "db.disableFreeMonitoring()"
```
If you have newer MongoDB shell and bash, you can also use in bash script these commands:
```
mongosh --eval "disableTelemetry()"

mongosh --eval "db.disableFreeMonitoring()"
```
I think I should include these commands to WeKan code.

When logging into interactive MongoDB shell, there is this complete text:

```
$ mongosh
Connecting to:		mongodb://127.0.0.1:27017/?directConnection=true&serverSelectionTimeoutMS=2000&appName=mongosh+1.8.0
Using MongoDB:		4.4.1
Using Mongosh:		1.8.0

For mongosh info see: https://docs.mongodb.com/mongodb-shell/

To help improve our products, anonymous usage data is collected and sent to MongoDB periodically (https://www.mongodb.com/legal/privacy-policy).
You can opt-out by running the disableTelemetry() command.

------
   The server generated these startup warnings when booting
   2023-03-13T09:17:36.967+02:00: ***** SERVER RESTARTED *****
   2023-03-13T09:17:37.069+02:00: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine. See http://dochub.mongodb.org/core/prodnotes-filesystem
   2023-03-13T09:17:37.274+02:00: Access control is not enabled for the database. Read and write access to data and configuration is unrestricted
   2023-03-13T09:17:37.274+02:00: You are running this process as the root user, which is not recommended
   2023-03-13T09:17:37.274+02:00: Soft rlimits too low
------
   Enable MongoDB's free cloud-based monitoring service, which will then receive and display
   metrics about your deployment (disk utilization, CPU, operation statistics, etc).
   
   The monitoring data will be available on a MongoDB website with a unique URL accessible to you
   and anyone you share the URL with. MongoDB may use this information to make product
   improvements and to suggest MongoDB products and deployment options to you.
   
   To enable free monitoring, run the following command: db.enableFreeMonitoring()
   To permanently disable this reminder, run the following command: db.disableFreeMonitoring()
------

test> 
```
Snap MongoDB is at port 27019, so connecting to it would be:
```
mongosh --port 27019
```

**Comment by xet7** [comment]
So after disabling those, you could try to remove apparmor rules, and see is there less noise in logs.

**Comment by webenefits** [comment]
Hi @xet7,

I finally got around to testing your suggestion. Unfortunately without success. 

1. First I installed the mongo shell 
```
wget https://downloads.mongodb.com/compass/mongodb-mongosh_1.8.0_amd64.deb
sudo apt install ./mongodb-mongosh_1.8.0_amd64.deb
```

2. Then I executed the suggested commands
```
mongosh --port 27019 --eval "disableTelemetry()"
mongosh --port 27019 --eval "db.disableFreeMonitoring()"
```

3. Then I undid the apparmor settings and compiled it again
```
sudo nano /var/lib/snapd/apparmor/profiles/snap.wekan.mongodb

sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.wekan.mongodb
sudo systemctl restart snap.wekan.mongodb.service snap.wekan.wekan.service
```

**Unfortunately there seems to be no change when I look at the syslog.**
Is there anything at all wrong with leaving the apparmor settings as they are?

**Comment by webenefits** [comment]
The cause was not found but with the workaround it works fine. Can probably be closed.

**Comment by diegargon** [comment]
That work for me (mongosh)

db.adminCommand( { setParameter: 1, diagnosticDataCollectionEnabled: false } )

**Comment by webenefits** [comment]
Oh yes, thank you. I can confirm that. After submitting the following command, the massive logs stopped.

```
mongosh --port 27019 --eval "db.adminCommand({setParameter: 1, diagnosticDataCollectionEnabled: false})"
```

Is this setting also retained after an update of wekan?

**Comment by xet7** [comment]
@diegargon @webenefits 

Please test did I disable apparmor denied messages in newest WeKan Snap Candidate correctly.

**Comment by webenefits** [comment]
Sorry, but unfortunately this did not work in my case. After updating to 7.19, the log entries appear again.

**Comment by xet7** [comment]
@webenefits 

Thanks! I'll look how to fix so that those commands are run correctly.

**Comment by diegargon** [comment]
same here

**Comment by chdcomputers** [comment]
> Thank you very much for your help.
> 
> Unfortunately, the [mentioned option mongo-log-destination »](https://github.com/wekan/wekan/issues/4855#issuecomment-1468768042) did not work. No matter whether `'snapcommon'` or `'devnull'`, the errors were written to the syslog. All other values resulted in the service not being able to start.
> 
> But with the Infos from https://snapcraft.io/docs/debug-snaps#heading--apparmor I edited the File '/var/lib/snapd/apparmor/profiles/snap.wekan.mongodb' and added following lines at the very end:
> 
> ```
> @{PROC}/@{pid}/net/netstat r,
> @{PROC}/@{pid}/mountinfo r,
> @{PROC}/@{pid}/net/snmp r,
> /proc/vmstat r,
> ```
> 
> Then, to compile, load the policy into the kernel and restart the service:
> 
> ```
> sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.wekan.mongodb
> sudo systemctl restart snap.wekan.mongodb.service snap.wekan.wekan.service
> ```
> 
> That seems to have fixed the problem. 🎉
> 
> However, as I did not understand 100% what exactly the error message originally said, it would be great if someone could check whether this is ok, especially with regard to security.

In addition to the lines you mention I had to add the following lines:

```
@{PROC}/pressure/io r,
@{PROC}/pressure/memory r,
@{PROC}/pressure/cpu r,
```
Still don't have a clue why this is happening though.

**Comment by xet7** [comment]
@webenefits 

Is this fixed in WeKan Snap Candidate v7.30 ?

**Comment by webenefits** [comment]
Hi @xet7 

unfortunately no, the bug/error still exists in v7.30. I have just made an update and had a look at the syslog

**Comment by webenefits** [comment]
Hi @xet7

the bug still exists in the current WeKan Snap Candidate v7.53. Is there a plan to fix this in the near future?

The problem probably occurs unnoticed by many and can lead to the syslog becoming extremely large, which in turn can lead to other errors.

**Comment by xet7** [comment]
@webenefits 

Did you set any different logging settings with snap commands?

You can see your current snap settings with:

```
sudo snap get wekan
```

**Comment by xet7** [comment]
@webenefits 

You see all possible WeKan Snap settings, including logging settings, with:
```
wekan.help | less
```

**Comment by xet7** [comment]
@webenefits 

It looks like default MongoDB log destination is `/dev/null`, so logs are not saved:

https://github.com/wekan/wekan/blob/main/snap-src/bin/config#L45

So maybe you could remove your setting, to have default setting:
```
sudo snap unset wekan default-mongo-log-destination
```
Or alternatively, set log destination to `/dev/null`:
```
sudo snap set wekan default-mongo-log-destination="devnull"
```
But, if these settings do not help, then maybe there is a bug somewhere.

**Comment by xet7** [comment]
I would presume, that this is already fixed, so I will close this issue.

If someone still has problems, please comment on this issue, so I will then open this issue again.

**Comment by webenefits** [comment]
Hi @xet7 ,

the problem still exists! Many people probably just don't notice it because the only symptom is a very, very large syslog. @chdcomputers @diegargon @mfilser what is your situation?

Our log destination is default. This has never been changed. Below is the output of `sudo snap get wekan`:

```
Key                    Value
debug                  false
mail-from              Wekan <***************>
mail-url               smtps://*******************************
mongo-log-destination  devnull
port                   3001
root-url               https://wekan****************
```

Can't you just include the code from my workaround in your Snap Candidate?

**Comment by mfilser** [comment]
@webenefits 
I don't use snap at all.

**Comment by xet7** [comment]
@webenefits 

> the only symptom is a very, very large syslog

Could you setup your system to rotate syslog daily?

**Comment by xet7** [comment]
@webenefits 

I'll look is there more settings for turning off MongoDB logs.

**Comment by xet7** [comment]
@webenefits 

I need to check, is that command already run correctly:

https://github.com/wekan/wekan/blob/main/snap-src/bin/mongodb-control#L119

**Comment by webenefits** [comment]
Ok :+1: 

What I just noticed is that the port parameter (`--port 27019`) is missing, but I don't know if it is needed at that place.

**Comment by xet7** [comment]
@webenefits 

There is `$SH_BIND_OPTIONS` variable at that line:

https://github.com/wekan/wekan/blob/main/snap-src/bin/mongodb-control#L119

That variable should have MongoDB port, if there is not some bug:

https://github.com/wekan/wekan/blob/main/snap-src/bin/mongodb-control#L93

**Comment by webenefits** [comment]
Ah ok right, I see. Yes then test it and give us feedback on the result, then we can check further.

**Comment by xet7** [comment]
It looks like Snap Candidate did not run that command by default. Spamming syslog stopped only after running that command. I'll try to get that command added and run by default, and make a new WeKan release.

**Comment by xet7** [comment]
@lorodoes 

I don't understand how these changes would help upgrading from Stable to Candidate. I try to revert those changes, maybe after that logs start working correctly.

https://github.com/wekan/wekan/blob/main/CHANGELOG.md#v741-2024-03-18-wekan--release

**Comment by webenefits** [comment]
Unfortunately, the bug is still present in version 7.56. :(

**Comment by xet7** [comment]
I need to try again. It did not work yet, Snap went to reboot loop. I try get those commands running correctly.

**Comment by xet7** [comment]
This does not work yet. Oh well. I'll try again later.

**Comment by mgdbbrt** [comment]
I think the fixes in 5aa38c do not work because the script is blocked until mongod is stopped.

**Comment by mgdbbrt** [comment]
One solution would be to start mongod in the background and wait for it later.

**Comment by xet7** [comment]
@mgdbbrt 

When running outside of WeKan Snap, this kind of script works, but I have not yet got it run from inside of WeKan Snap.

```
# Disable telemetry

/snap/wekan/current/usr/bin/mongosh wekan --eval 'disableTelemetry();' --port 27019

/snap/wekan/current/usr/bin/mongosh wekan --eval 'db.adminCommand({ setParameter: 1, diagnosticDataCollectionEnabled: false});' --port 27019

# Delete incomplete uploads so that they would not prevent starting WeKan

/snap/wekan/current/usr/bin/mongosh wekan --eval 'db.getCollection("cfs.attachments.filerecord").find( { "uploadedAt": { "$exists": true }, "copies.attachments" : null,"failures.copies.attachments.doneTrying" : {"$ne" : true}});' --port 27019
```

**Comment by stefanpuch** [comment]
Unfortunately, the bug is still present in version 7.88 (Rev 3103 latest/candidate)

Since I did not install Mongodb tools on Ubuntu 24.04 so far I tested the workaround from  https://github.com/wekan/wekan/issues/4855#issuecomment-1853071612

I edited the File '/var/lib/snapd/apparmor/profiles/snap.wekan.mongodb' and added following seven lines at the very end:
```
@{PROC}/@{pid}/net/netstat` r,
@{PROC}/@{pid}/mountinfo r,
@{PROC}/@{pid}/net/snmp r,
/proc/vmstat r,
@{PROC}/pressure/io r,
@{PROC}/pressure/memory r,
@{PROC}/pressure/cpu r,
```

After that I compiled, and load the policy into the kernel and restart the service:
``` 
sudo apparmor_parser -r /var/lib/snapd/apparmor/profiles/snap.wekan.mongodb
sudo systemctl restart snap.wekan.mongodb.service snap.wekan.wekan.service
``` 

That stopped the messages in syslog but I assume that they will return if wekan snap is updated the next time since then file will be replaced right?

**Comment by xet7** [comment]
@stefanpuch 

Yes, that's right. Some additional commands are here:

https://github.com/wekan/wekan/wiki/Cron

I have tried to get those commands to run inside of WeKan Snap, but it did not work yet.

I try to sometime get working migrating WeKan Snap from MongoDB to SQLite. Then there would be no logs generated at all.

**Comment by xet7** [comment]
Also, that restart cron is because currently there is memory leaks in WeKan, memory and CPU usage keeps increasing, so WeKan may freeze and need restarting once a week. According to VSCode Copilot, that could be because of memory leaks at reactiveCache https://github.com/wekan/wekan/pull/5014 , but I have not yet tried to fix it, because I'm also trying to do many other upgrades.

**Comment by xet7** [comment]
If there is problems with MongoDB, try instead FerretDB 2 / PostgreSQL:

https://github.com/wekan/wekan/blob/main/docs/Databases/FerretDB2-PostgreSQL.md
